### PR TITLE
Changes to enable Nitro enclaves

### PR DESCRIPTION
**Pull request by dreemkiller**
Added a new docker target nitro. 
Certain things needed to be done inside the make file (like building the AWS nitro-cli tools) instead of the docker file as I would have liked.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -13,53 +13,26 @@
...
+FROM ubuntu:18.04 AS build_root
 ARG USER
 ARG UID
-ENV DEBIAN_FRONTEND noninteractive
-# Use bash as the default
-SHELL ["/bin/bash", "-c"]
+
+# add a user
+RUN mkdir -p /work
+RUN mkdir -p /poop
```
Comment (ericvh) [inline Dockerfile]: nice

File: Dockerfile [hunk 2]
```diff
@@ -13,53 +13,26 @@
...
+RUN if [ "$USER" != "root" ] ; then useradd -u $UID -m -p deadbeef -s /bin/bash $USER ; fi; \
```
Comment (ericvh) [inline Dockerfile]: I've got a slightly more secure variation of this I use:
        if [ "$USER" != "root" ] ; then useradd -u $UID -m -p `openssl rand -base64 32` -s /bin/bash $USER ; fi; \

File: Dockerfile [hunk 3]
```diff
@@ -311,10 +314,72 @@ RUN chown -R $USER /work/rust-optee-trustzone-sdk/optee-qemuv8-3.7.0/build/build
 WORKDIR /
 RUN ./cleanup.sh
 
+
```
Comment (ericvh) [inline Dockerfile]: Am I reading this right?  Are you always installing the nitro stuff in base?  Should we isolate it?

File: Makefile [hunk 4]
```diff
@@ -37,9 +37,26 @@ else # otherwise linux
 	docker run --privileged -e DISPLAY=${DISPLAY} -d -v /tmp/.X11-unix:/tmp/.X11-unix -v $(abspath $(VERACRUZ_ROOT)):/work/veracruz --name  $(VERACRUZ_CONTAINER) $(VERACRUZ_DOCKER_IMAGE)_tz
 endif
 
+.PHONY:
+nitro: build
+ifeq ($(OS_NAME),darwin)
+	docker run --privileged -e DISPLAY=$(IP):0 -d -v /tmp/.X11-unix:/tmp/.X11-unix -v $(abspath $(VERACRUZ_ROOT)):/work/veracruz -v $(HOME)/.cargo/registry/:/usr/local/cargo/registry/ --name  $(VERACRUZ_CONTAINER) $(VERACRUZ_DOCKER_IMAGE)_nitro
```
Comment (ericvh) [inline Makefile]: does nitro really need the X11 stuff?

File: Dockerfile (hunk 1 above)
Comment (dreemkiller) [reply Dockerfile]: That was a test. You passed.

File: Dockerfile (hunk 3 above)
Comment (dreemkiller) [reply Dockerfile]: probably so. I'm going to do a rebase on top of your changes, and I'll move it.

File: Makefile (hunk 4 above)
Comment (dreemkiller) [reply Makefile]: It does not.

**Comment by ericvh** [comment]
rebase is going to be a bit odd, since what i did was split the target
specific docker files to their own subdirs -- so basically I 'accepted all'
in the main Dockerfile and then just moved the new stuff to a subdir.

**Comment by dreemkiller** [comment]
Yeah, I was using the term "rebase" loosely. More like "rewrite everything on top of the latest version".

**Comment by dreemkiller** [comment]
Yeah, I was using the term "rebase" loosely. More like "rewrite everything on top of the latest version".

That's done now.

File: Makefile [hunk 5]
```diff
@@ -55,6 +64,18 @@ sgx-base: sgx/Dockerfile base
...
 	DOCKER_BUILDKIT=1 docker build --build-arg USER=root --build-arg UID=0 --build-arg TEE=tz -t veracruz/tz-base -f $< .
 
+.PHONY:
+nitro-base: nitro/Dockerfile base
+ifneq (,$(wildcard aws-nitro-enclaves-cli))
+	cd "aws-nitro-enclaves-cli"
+	git pull origin main
+else
+	git clone https://github.com/aws/aws-nitro-enclaves-cli.git
+endif
```
Comment (nspin) [inline Makefile]: Pinning a revision or version branch could make this part more reproducible. Not necessary of course but we have talked about striving for reproducibility.

(I'm talking about reproducibility in the informal sense, not the bit-for-bit sort we've also discussed)

Comment (dreemkiller) [reply Makefile]: That's a good idea. Will do.

Comment (dreemkiller) [reply Makefile]: I've done this now. Take a look

**Comment by dreemkiller** [comment]
@nspin Is there anything holding this up now?

**Comment by dreemkiller** [comment]
Can we get this approved? @ShaleXIONG ? It's been sitting here a while.

**Comment by dreemkiller** [comment]
@DominicPM Do you mind reviewing this? I cant' get any action on it.

**Comment by dominic-mulligan-arm** [comment]
Sure, I'll take a look.

File: Makefile [hunk 6]
```diff
@@ -5,6 +5,8 @@ USER ?= $(shell id -un)
 UID ?= $(shell id -u)
 IP := $(firstword $(shell ip addr show | grep -Eo 'inet (addr:)?([0-9]*\.){3}[0-9]*' | grep -Eo '([0-9]*\.){3}[0-9]*' | grep -v '127.0.0.1' | awk '{print $1}' ))
 OS_NAME := $(shell uname -s | tr A-Z a-z)
+LOCALIP=$(shell ip -4 address show eth0 | grep -Eo 'inet (addr:)?([0-9]*\.){3}[0-9]*' | grep -Eo '([0-9]*\.){3}[0-9]*')
+AWS_NITRO_CLI_REVISION = dff9102783959412dcf5d515e641c2c3ad0d443b
```
Comment (dominic-mulligan-arm) [inline Makefile:9]: Is this particular revision important?

**Comment by dominic-mulligan-arm** [comment]
Skimmed through manually and only have one comment.  I'll checkout locally and try to build, too.

**Review by ShaleXIONG — APPROVED** [review]

**Comment by dominic-mulligan-arm** [comment]
@ShaleXIONG did you check that this builds correctly?  I seem to remember you did.  If so, I don't need to do it and this can be merged.

**Comment by ShaleXIONG** [comment]
> @ShaleXIONG did you check that this builds correctly? I seem to remember you did. If so, I don't need to do it and this can be merged.

Yes, I checked it a long time ago, forgot to approve.

**Review by nspin — APPROVED** [review]

File: Makefile [hunk 7]
```diff
@@ -47,6 +49,14 @@ endif
 ci-exec:
 	docker exec -i -t $(VERACRUZ_CONTAINER)_ci_$(USER) /bin/bash
 
+.PHONY:
+nitro-run: nitro-base
+	docker run --privileged -d -v $(abspath $(VERACRUZ_ROOT)):/work/veracruz -v $(HOME)/.cargo/registry/:/usr/local/cargo/registry/ -v /usr/bin:/host/bin -v /usr/share/nitro_enclaves:/usr/share/nitro_enclaves -v /run/nitro_enclaves:/run/nitro_enclaves -v /etc/nitro_enclaves:/etc/nitro_enclaves --device=/dev/vsock:/dev/vsock -v /var/run/docker.sock:/var/run/docker.sock --device=/dev/nitro_enclaves:/dev/nitro_enclaves --env TABASCO_IP_ADDRESS=$(LOCALIP) -p $(LOCALIP):3010:3010/tcp --name $(VERACRUZ_CONTAINER)_nitro_$(USER) $(VERACRUZ_DOCKER_IMAGE)_nitro:$(USER) sleep inf
```
Comment (dominic-mulligan-arm) [inline Makefile:54]: A stray mention of Tabasco in `TABASCO_IP_ADDRESS`.

Comment (dominic-mulligan-arm) [reply Makefile:54]: Is that even needed any more?

File: Makefile (hunk 6 above)
Comment (dreemkiller) [reply Makefile:9]: Not particularly. But I added the specific version at @nspin 's suggestion to make builds repeatable (which is a Good Idea).

Comment (dominic-mulligan-arm) [reply Makefile:9]: OK, that makes sense.

File: Makefile (hunk 7 above)
Comment (dreemkiller) [reply Makefile:54]: Yes, still needed until #180 is merged. The name is anachronistic, but exists in current code on main until the #180 merge.

Comment (dominic-mulligan-arm) [reply Makefile:54]: OK, cool.

**Review by dominic-mulligan-arm — APPROVED** [review]

**Comment by dominic-mulligan-arm** [comment]
+1+1+1 = +1 reached.  merging.

**Pull request merged**